### DISCUSSION TITLE
translation (i18) / localisation (l10n) support

### DISCUSSION THREAD

**Discussion by amitu** [fastn-stack/fastn#1591] [Ideas & RFCs]
- [x] basic `lang` global 
- [ ] `<package-name>/lang` so any package can use any other package's lang stuff
- [ ] `<package-name>/lang/<foo>` imports should also work

<details>
<summary>Changes</summary>

- AmitU, 12th Dec 2023: changed `i18n` to `lang`, as `lang` is auto import alias, and also `lang` handles both `i18n` and `l10n` concerns. 

</details>

# All Translations In Single Package

```ftd
-- import: fastn
-- fastn.package: slides.fifthtry.com
default-lang: en
lang: slides.fifthtry.com/my-lang-files
```

We are introducing one key `lang`, which must be the folder in which translation files are present, eg in this case we have `my-lang-files` folder, which should contain files like `en.ftd` (or `en-us.ftd`, `en-in.ftd`), `hi.ftd` and so on. 

# Each Translation In Individual Package

```ftd
-- import: fastn
-- fastn.package: slides.fifthtry.com
lang-en: slides.fifthtry.com/my-en-strings
lang-hi: slides-hi.fifthtry.com

-- fastn.dependency: slides-hi.fifthtry.com
```

You can mix and match also:

```ftd
-- import: fastn
-- fastn.package: slides.fifthtry.com
lang: slides.fifthtry.com/my-i18n-files
lang-hi: slides-hi.fifthtry.com

-- fastn.dependency: slides-hi.fifthtry.com
```

In this setup, hi translation will be found in `lang-hi` module, and all other languages will be looked in `my-lang-files` folder (if `hi.ftd` is present in `my-lang-files`, then this would be a compilation error).

# All modules must match

Basically it is a compilation error if some key is present in one language file but not in another. 

# Localisation Helpers

`ftd.lang.hi.number_to_string()` etc will localise a number to string.

# Language Detection / Activation

`fastn` will honour a cookie `fastn_lang`, with the language that user has currently selected. If cookie is missing we will look at request header and figure out the best language match. 

# `lang` "system"

Any package that has language configured using `lang` keys, will have access to `lang` auto imported module, which will point to the language module for the current request. 

# Language Files Do Not Just Contain Strings

They can contain functions, `ftd.image-src` or even components.

# This Is Not Content Translation

This feature can be used for content translation, but it is largely meant for translating fastn powered user interfaces. 

Content translation imposes additional requirements, like each language must have a searchable permanent URL. This scheme uses same URL for content, and uses cookie/accept-lang header to switch language. If you are writing a blog or a book and want it translated in multiple language, wait for the content translation feature.

**Comment by Arpita-Jaiswal** [comment]
The default language should be decided by top-level package, if that language is not available then fallback to default of dependency package.
